### PR TITLE
Disable Local Echo

### PR DESCRIPTION
**Pull request by chenqiang1986**

### CODE REVIEW

**Comment by dontcallmedom** [comment]
@chenqiang1986 I understand you work for Google; would you consider [requesting to be added to the Working Group](https://www.w3.org/2004/01/pp-impl/47318/join)?

**Comment by chenqiang1986** [comment]
Any further action? How can I link W3C account with github account?

**Comment by plehegar** [comment]
@chenqiang1986 you're all set. Your account has been linked (it's done manually so far unfortunately).

**Comment by chenqiang1986** [comment]
Whom should I notify for the review?

**Comment by martinthomson** [comment]
Discussion of this spec happens primarily on public-media-capture@w3.org.  If you want to start a thread there, that would be a good way to start.

**Comment by chenqiang1986** [comment]
I started a thread in public-media-capture@w3.org months ago, but no reply at all.
Anything else I can do?

**Comment by martinthomson** [comment]
Maybe you could convince the group that this is worth doing.  I don't think that this is a good idea, but I'm always open to arguments about why something might be useful.

**Comment by alvestrand** [comment]
The next question is "what's your use case". The mail to the TF list was better at explaining it than this issue itself. Do you want to repeat it here?

File: index.html [hunk 1]
```diff
@@ -201,6 +204,15 @@ <h2 id="constraints">Constraining Display Surface Selection</h2>
         the <a>visible display surface</a>.</p>
 
         <dl class="idl" title="partial dictionary MediaTrackConstraintSet">
+          <dt>ConstrainBoolean disableLocalEcho</dt>
+
+          <dd>
+            <p>A value of <code>true</code> indicates that we mute the local
+            playback for the captured content. For example, when capturing a
+            browser tab, the tab will be muted then; when capturing the full
+            screen with audio, then the system audio will be muted then.</p>
```
Comment (alvestrand) [inline index.html]: Grammar nit: I think you can delete all occurences of "then" in this sentence and make it more readable.

File: index.html [hunk 2]
```diff
@@ -186,11 +186,14 @@ <h2 id="constraints">Constraining Display Surface Selection</h2>
...
       <section>
         <h2>New Constraints for Captured Display Surfaces</h2>
-        <p>Two constraints are defined that allow an application to observe
+        <p>Three constraints are defined that allow an application to observe
         properties of the selected <a>display surface</a>.  Since the source of
         media cannot be changed after a <code>MediaStreamTrack</code> has been
         returned, these constraints cannot be changed by an application.</p>
 
+        <p>The <code><a href="#widl-MediaTrackConstraintSet-disableLocalEcho">disableLocalEcho</a></code>
+        constraint allows an application to suppress local playback of the captured audio.</p>
```
Comment (alvestrand) [inline index.html]: A name of "muteAudioForSource" might be easier to understand. Having booleans with "disable" in the name is usually bad practice.

File: index.html [hunk 3]
```diff
@@ -186,11 +186,14 @@ <h2 id="constraints">Constraining Display Surface Selection</h2>
 
       <section>
         <h2>New Constraints for Captured Display Surfaces</h2>
-        <p>Two constraints are defined that allow an application to observe
+        <p>Three constraints are defined that allow an application to observe
         properties of the selected <a>display surface</a>.  Since the source of
         media cannot be changed after a <code>MediaStreamTrack</code> has been
         returned, these constraints cannot be changed by an application.</p>
```
Comment (alvestrand) [inline index.html:193]: This description seems not accurate any more, since the new constraint modifies a property of the captured surface. Can you rephrase?

**Review by alvestrand — COMMENTED** [review]
This doesn't seem an unreasonable thing to ask for, but it's not completely done yet.
You will also want to mention what you expect the normal behavior (no constraint) to be - I'd expect it to be that the captured surface's associated sound continues to play locally - that is, "disableLocalAudio=false".

**Comment by chenqiang1986** [comment]
We are actually driven by the use cases from users, as we found without such freedom, either default setting (mute or unmute), some users will be unhappy.

1. For peer to peer connection, for example hangouts session or any other video chat, when sharing audio, the user would in general expect that both the sender and receiver can hear the sound.

2. For Cast, or any other app that is designed to project a laptop to the big screen in the conference room, we would hope that only the receiver (the speaker associated with the big screen) would play out the audio.

**Comment by chenqiang1986** [comment]
Thanks for review. Rephrased.

File: index.html [hunk 4]
```diff
@@ -217,6 +221,15 @@ <h2>New Constraints for Captured Display Surfaces</h2>
...
           </dd>
+
+          <dt>ConstrainBoolean muteAudioForSource</dt>
+
+          <dd>
+            <p>A value of <code>true</code> indicates that we mute the local
+            playback for the captured content. For example, when capturing a
+            browser tab, the tab will be muted; when capturing the full screen
+            with audio, then the system audio will be muted.</p>
+          </dd>
```
Comment (henbos) [inline index.html:232]: For tab capture, does "The tab will be muted" mean everything is silence, or does this mean we get all system audio except for that tab?

When capturing the full screen, is muting the system audio the same as not requesting audio?

What do you think about this proposal?
https://docs.google.com/document/d/1pgN8Pe2zMaD0tytJVsKmnBT-7gM247YlzzWazm7OwSk/edit?usp=sharing
It overlaps with some of this.

**Comment by henbos** [comment]
I filed #79

**Comment by alvestrand** [comment]
Closing as OBE.
